### PR TITLE
Issue #85 Solution (for HTTP Protocol)

### PR DESCRIPTION
**Pull request by DeveshChande**
heralding can now log : 
- User Agent
- Which version of python is being used
- Which version of HTTP (Python) is being used

### CODE REVIEW

**Comment by DeveshChande** [comment]
In bin/heralding, the class "LogFilter" has a function called filter. Comparing with python documentation about the logging module [here](https://docs.python.org/2/library/logging.html#logging.Filter.filter), it seems clear the "self" attribute is not required. Moreover, it looks like heralding is not using it as well (in the filter function).

File: heralding/capabilities/http.py [hunk 1]
```diff
@@ -66,6 +73,11 @@ def do_GET(self):
             self.wfile.write(headers_bytes)
             self.wfile.write(b'not authenticated')
 
+        logger.debug(self.headers['User-Agent'])
```
Comment (johnnykv) [inline heralding/capabilities/http.py]: Did you study how we log data in Heralding? There are two "types" of logging, one if stuff regarding Heralding itself the other is the stuff we log about the clients that try to authenticate. Logging the User agent would be the latter, hence this is not the proper way to do this. Try studying the code in the Reporting directory.

**Comment by johnnykv** [comment]
The travis test cases on 3.5.0 fails.

**Comment by kajojify** [comment]
Hi there! :) 
Congratulations with your first PR for Heralding :tada:
Thank you for taking this. 

1. You probably didn't want to rename _bin/heralding_ to _bin/heralding.py_
Read this - https://stackoverflow.com/a/5998845 and this - https://coderwall.com/p/lt2kew/python-creating-your-project-structure.  
There was an intention to write _bin/heralding_.

2. You put new option --stat. I don't see where you are reading this value? 

3. Let's talk about the structure a bit. --stat option is good, but I think we need a bit different approach.
We will probably want to have some statistics about ssh for example or about any other protocol. We also don't want to collect all information about all protocols each time (if a user doesn't want), so we have to provide the possibility to choose protocols for grabbing statistics. --stat won't suit, since I think it will be difficult to choose protocols for grabbing additional information from the command line.
We have _heralding/heralding.yaml_ for such purposes. So, I think the possible way would be to add specific options for grabbing into _heralding.yaml_ and then turn stats on or off for specific protocols from the code depending on the option values.

**Comment by kajojify** [comment]
4. Could you please provide any use case, why we need to log python version and server version after each request? I don't see how python version and server version might be changed after the launch, so they will be the same for each request. I think if we want to log this, we should do it inside the main log file (heralding.log) and only once.
5. Code with logger creation should go to _Reporting directory_. 
6. > In bin/heralding, the class "LogFilter" has a function called filter. Comparing with python documentation about the logging module [here](https://docs.python.org/2/library/logging.html#logging.Filter.filter), it seems that the "self" attribute is not required. Moreover, it looks like heralding is not using it as well (in the filter function).

a) Use the docs for Python 3, nor for Python 2 :)  
https://docs.python.org/3/library/logging.html#logging.Filter
b) _self_ is omitted almost everywhere in the docs. So if _self_ is absent, this doesn't mean it isn't used. 
c) look at the source code of _logging.Filter_: https://github.com/python/cpython/blob/3.7/Lib/logging/__init__.py#L707-L720
_filter_ uses _self_, so we are overriding it with the same signature

**Comment by DeveshChande** [comment]
Thank you @johnnykv and @kajojify for the guidance. I will get to work.

**Comment by DeveshChande** [comment]
> 1. Could you please provide any use case, why we need to log python version and server version after each request? I don't see how python version and server version might be changed after the launch, so they will be the same for each request. I think if we want to log this, we should do it inside the main log file (heralding.log) and only once.
> 2. Code with logger creation should go to _Reporting directory_.
> 3. > In bin/heralding, the class "LogFilter" has a function called filter. Comparing with python documentation about the logging module [here](https://docs.python.org/2/library/logging.html#logging.Filter.filter), it seems that the "self" attribute is not required. Moreover, it looks like heralding is not using it as well (in the filter function).
> 
> a) Use the docs for Python 3, nor for Python 2 :)
> https://docs.python.org/3/library/logging.html#logging.Filter
> b) _self_ is omitted almost everywhere in the docs. So if _self_ is absent, this doesn't mean it isn't used.
> c) look at the source code of _logging.Filter_: https://github.com/python/cpython/blob/3.7/Lib/logging/__init__.py#L707-L720
> _filter_ uses _self_, so we are overriding it with the same signature

With reference to the 3rd point, logging everything about every protocol does seem to be unnecessary. 


To answer the 4th point, you are right. There is no use of logging python version and server version each time a request is made. Main motive behind logging it was to check attribute values without running commands in terminal ( for example :  python --version). Logging it once and moving on is much more appropriate.

**Comment by johnnykv** [comment]
I am not sure this is the correct way to to this. Each protocol (capability) might want to log auxiliary data and we need a somewhat uniform way to do this. Some solutions to this might be:
* One global log for auxiliary data. Each log element would be a XML node. This allows for different formats for each line.
* One auxiliary log file for each protocol, could be CSV for XML.

**Comment by johnnykv** [comment]
I am not sure this is the correct way to to this. Each protocol (capability) might want to log auxiliary data and we need a somewhat uniform way to do this. Some solutions to this might be:

* One global log for auxiliary data. Each log element would be a XML node. This allows for different formats for each line.
* One auxiliary log file for each protocol, could be CSV for XML.

Before any more code is submitted, I would like to see a overall plan for logging auxiliary data.

**Pull request closed without merging**